numpy/lib/arraypad.py

```diff
@@ -1211,174 +809,108 @@ def pad(array, pad_width, mode, **kwargs):
-            newmat = _append_med(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'minimum':
-        for axis, ((pad_before, pad_after), (chunk_before, chunk_after)) \
-                in enumerate(zip(pad_width, kwargs['stat_length'])):
-            newmat = _prepend_min(newmat, pad_before, chunk_before, axis)
-            newmat = _append_min(newmat, pad_after, chunk_after, axis)
-
-    elif mode == 'reflect':
-        for axis, (pad_before, pad_after) in enumerate(pad_width):
-            if narray.shape[axis] == 0:
+    stat_functions = {
+        "maximum": np.max,
+        "minimum": np.min,
+        "mean": np.mean,
+        "median": np.median,
+    }
+
+    # Make sure that no unsupported keywords were passed for the current mode
+    unsupported_kwargs = set(kwargs) - set(allowed_kwargs[mode])
+    if unsupported_kwargs:
+        raise ValueError("unsupported keyword arguments for mode '{}': {}"
+                         .format(mode, unsupported_kwargs))
+
+    # Set kwarg defaults
+    for key in allowed_kwargs[mode]:
+        kwargs.setdefault(key, kwarg_defaults[key])
+
+    # Create array with final shape and original values
+    # (padded area is undefined)
+    padded, _ = _pad_simple(array, pad_width)
+    # And prepare iteration over all dimensions
+    # (zipping may be more readable than using enumerate)
+    axes = range(padded.ndim)
+
+    if mode == "constant":
+        values = _as_pairs(kwargs["constant_values"], padded.ndim)
+        for axis, index_pair, value_pair in zip(axes, pad_width, values):
+            _set_pad_area(padded, axis, index_pair, value_pair)
+
+    elif mode == "edge":
+        for axis, index_pair in zip(axes, pad_width):
+            edge_pair = _get_edges(padded, axis, index_pair)
+            _set_pad_area(padded, axis, index_pair, edge_pair)
+
+    elif mode == "linear_ramp":
+        end_values = _as_pairs(kwargs["end_values"], padded.ndim)
+        for axis, index_pair, value_pair in zip(axes, pad_width, end_values):
+            ramp_pair = _get_linear_ramps(padded, axis, index_pair, value_pair)
+            _set_pad_area(padded, axis, index_pair, ramp_pair)
+
+    elif mode in stat_functions.keys():
+        func = stat_functions[mode]
+        length = _as_pairs(kwargs["stat_length"], padded.ndim, as_index=True)
+        for axis, index_pair, length_pair in zip(axes, pad_width, length):
+            stat_pair = _get_stats(padded, axis, index_pair, length_pair, func)
+            _set_pad_area(padded, axis, index_pair, stat_pair)
+
+    elif mode in {"reflect", "symmetric"}:
+        method = kwargs["reflect_type"]
+        include_edge = True if mode == "symmetric" else False
+        for axis, (left_index, right_index) in zip(axes, pad_width):
+            if array.shape[axis] == 0:
                 # Axes with non-zero padding cannot be empty.
-                if pad_before > 0 or pad_after > 0:
+                if left_index > 0 or right_index > 0:
```

Comment:
This case didn't use to error for `symmetric`

Reply:
Right, good catch! I'll add `and mode == "reflect"` to the if-statement.

Reply:
> This case didn't use to error for symmetric

I just noticed that modes "symmetric" and "wrap" fail in the old implementation anyway for this case. This seems like an oversight and I'm a little bit confused by the test `array.shape[axis] == 0`. If any axis in an array is 0 it's an empty array whose axes are always 0, correct? 
Furthermore the modes "reflect", "symmetric", "wrap" (stat modes as well) all derive the padded values from the original values and should fail gracefully which only "reflect" currently does. 

> If the test only appears in this PR, then I have no way of knowing if the behavior is unchanged

Yeah, making another PR with additional tests is no problem. But what about tests, that are expected to fail for the old implementation? E.g. the regression test for #11216 or testing "symmetric" / "wrap" with empty arrays? Should I add them marked with `xfail`?

> Since this is looking like a total rewrite, it would be good to have it as tested as possible first.

Yeah sorry, about not being clear about that upfront. That wasn't my initial intention but at some point I decided that it would be easier to start with a blank slate and take the useful bits and ideas from the old implementation, of which there were many. And just to say it, the unit tests were extremely helpful in getting this up and running.

So I'll agree wholeheartedly about the testing part.

Reply:
> Should I add them marked with xfail?

That's probably what I'd do, but I'm not sure there's precedent for that

> If any axis in an array is 0 it's an empty array whose axes are always 0, correct?

Yes, but it should be perfectly legal for me to pad a `(10, 0)` array into `(20, 0)` by any mode - I think the current `pad` handles that correctly. Issues only arise when trying to extend the empty access, because now data has to be invented. In some cases (constant / empty), that invention is easy, but probably not for `reflect`